analyzers/PapasSim.py

```diff
@@ -56,84 +48,31 @@ def process(self, event):
         #random.seed(0xdeadbeef) #Useful to make results reproducable between loops and single runs
         event.simulator = self
         papasevent = PapasEvent(event.iEv)
+        if event.iEv ==60:
+            print "132"
         setattr(event, "papasevent", papasevent)        
-        pfsim_particles = []
         gen_particles = getattr(event, self.cfg_ana.gen_particles)
         try:
-            self.simulator.simulate(gen_particles)
+            self.simulator.simulate(gen_particles, papasevent.history)
         except (PropagationError, SimulationError) as err:
             self.mainLogger.error(str(err) + ' -> Event discarded')
             return False
-        pfsim_particles = self.simulator.ptcs
-
         #these are the particles before simulation
-        simparticles = sorted(pfsim_particles,
-                              key=lambda ptc: ptc.e(), reverse=True)     
+        simparticles = sorted(self.simulator.ptcs, key=P4.sort_key, reverse=True)
         setattr(event, self.simname, simparticles)
-    
-        #create dicts of clusters, particles etc (todo?:move a lot of this into simulator)
-        self.build_collections_and_history(papasevent, simparticles)
+        papasevent.history = self.simulator.history
+        papasevent.add_collection(self.simulator.simulated_particles)
```

Comment:
This would be good to discuss! 
What I have done here mirrors C++ (more or less). In C++ I wanted the possibility of the papasevent argument being a const object. This was motivated by Gaudi considerations ( at least for the initial gaudi implementations). There may well be a better way.